kin/monitors.py

```diff
@@ -1,227 +1,68 @@
 """Contains classes for monitoring the blockchain"""
-from threading import Thread, Event
-
 from .blockchain.utils import is_valid_address
-from .transactions import OperationTypes, SimplifiedTransaction ,RawTransaction
-from .errors import AccountNotFoundError, CantSimplifyError, StoppedMonitorError, StellarAddressInvalidError
+from .transactions import OperationTypes, SimplifiedTransaction, RawTransaction
+from .errors import CantSimplifyError, StellarAddressInvalidError
+
+from typing import Optional, AsyncGenerator
 
 import logging
 
 logger = logging.getLogger(__name__)
 
 
-class SingleMonitor:
-    """Single Monitor to monitor Kin payment on a single account"""
-
-    def __init__(self, kin_client, address, callback_fn):
-        """
-        Monitors a single account for kin payments
-        :param kin.KinClient kin_client: a kin client directed to the correct network
-        :param str address: address to watch
-        :param callback_fn: the function to call on each received payment as `callback_fn(address, tx_data, monitor)`.
-        :type: callable[str,kin.transactions.SimplifiedTransaction ,kin.monitors.MultiMonitor]
-        """
-        self.kin_client = kin_client
-        self.callback_fn = callback_fn
-
-        if not address:
-            raise ValueError('no address to monitor')
-
-        if not is_valid_address(address):
-            raise StellarAddressInvalidError('invalid address: {}'.format(address))
-
-        if not self.kin_client.does_account_exists(address):
-            raise AccountNotFoundError(address)
-
-        self.address = address
-
-        # Currently, due to nonstandard SSE implementation in Horizon,
-        # using cursor=now will block until the first tx happens.
-        # Instead, we determine the cursor ourselves.
-        # Fix will be for horizon to send any message just to start a connection.
-        # This will cause a tx
-        params = {}
-        reply = self.kin_client.horizon.account_transactions(address, params={'order': 'desc', 'limit': 2})
-        if len(reply['_embedded']['records']) == 2:
-            cursor = reply['_embedded']['records'][1]['paging_token']
-            params = {'cursor': cursor}
-
-        # make synchronous SSE request (will raise errors in the current thread)
-        self.sse_client = self.kin_client.horizon.account_transactions(address, sse=True, params=params)
-
-        self.stop_event = Event()
-        # start monitoring thread
-        self.thread = Thread(target=self.event_processor, args=(self.stop_event,))
-        self.thread.daemon = True
-        self.thread.start()
-
-    def event_processor(self, stop_event):
-        """
-        Method to filter through SSE events and find kin payments for an account
-        :param threading.Event stop_event: an event that can be used to stop this method
-        """
-        import json
+async def single_monitor(kin_client: 'KinClient', address: str,
```

Comment:
why does the KinClinet is wrapped with ' ?

Reply:
Kin client imports "Single Monitor"

If I wanted to write KinClient without '' I would need to import KinClient in SingleMonitor, and then I would have a circular loop of importing. So instead of importing the type you can just write it as a string. That's called 'forward referencing'

https://stackoverflow.com/questions/39740632/python-type-hinting-without-cyclic-imports

Reply:
python is fun!